File: 1034-subarrays-with-k-different-integers/subarrays-with-k-different-integers.java
class Solution {
    public int subarraysWithKDistinct(int[] nums, int k) {
        return atmost(nums,k)-atmost(nums,k-1);
    }
    private int atmost(int[] nums,int k){
        int left=0,count=0;
        HashMap<Integer,Integer> map= new HashMap<>();
        for(int right=0;right<nums.length;right++){
            map.put(nums[right],map.getOrDefault(nums[right],0)+1);
            while(map.size()>k){
                int l=nums[left];
                map.put(l,map.get(l)-1);
                if(map.get(l)==0){
                    map.remove(l);
                }
                left++;
            }
            count+=right-left+1;
        }
        return count;
    }
}

File: 1034-subarrays-with-k-different-integers/subarrays-with-k-different-integers.cpp
class Solution {
public:
    int solve(vector<int> &nums,int k){
        int n=nums.size();
        int ans=0;
        int l=0,r=0;
        map<int,int> mp;
        while(r<n){
            mp[nums[r]]++;
            while(mp.size()>k){
                mp[nums[l]]--;
                if(mp[nums[l]]==0){
                    mp.erase(nums[l]);
                }
                l++;
            }
            ans+=r-l+1;
            r++;
        }
        return ans;
    }
    int subarraysWithKDistinct(vector<int>& nums, int k) {
        int atmostk=solve(nums,k);
        int atmostk1=solve(nums,k-1);
        return atmostk-atmostk1;
    }
};